azure-kusto-data/source/tokenProvider.ts

```diff
@@ -0,0 +1,313 @@
+import { DeviceCodeResponse } from "@azure/msal-common";
+import { AzureCliCredentials } from "@azure/ms-rest-nodeauth";
+import { ManagedIdentityCredential } from "@azure/identity";
+import { CloudSettings, CloudInfo } from "./cloudSettings"
+
+export declare type TokenResponse = {
+    tokenType: string;
+    accessToken: string;
+}
+
+const BEARER_TYPE = "Bearer";
+
+/**
+ * This base class abstracts token acquisition for all implementations.
+ * The class is build for Lazy initialization, so that the first call, take on instantiation of 'heavy' long-lived class members
+ */
+export abstract class TokenProviderBase {
+    kustoUri: string;
+    scopes!: string[];
+
+    abstract name: string;
+    abstract acquireToken(): Promise<TokenResponse>;
+
+    constructor(kustoUri: string) {
+        this.kustoUri = kustoUri;
+        if (kustoUri != null) {
+            const suffix = this.kustoUri.endsWith("/") ? ".default" : "/.default";
+            this.scopes = [kustoUri + suffix];
+        }
+    }
+}
+
+/**
+ * Basic Token Provider keeps and returns a token received on construction
+ */
+export class BasicTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    token: string;
+
+    constructor(kustoUri: string, token: string) {
+        super(kustoUri);
+        this.token = token;
+    }
+
+    acquireToken(): Promise<TokenResponse> {
+        return Promise.resolve<TokenResponse>({ tokenType: BEARER_TYPE, accessToken: this.token });
+    }
+}
+
+/**
+ * Callback Token Provider generates a token based on a callback function provided by the caller
+ */
+export class CallbackTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    callback: () => Promise<string>;
+
+    constructor(kustoUri: string, callback: () => Promise<string>) {
+        super(kustoUri);
+        this.callback = callback;
+    }
+
+    async acquireToken(): Promise<TokenResponse> {
+        const token = await this.callback();
+        return { tokenType: BEARER_TYPE, accessToken: token };
+    }
+}
+
+
+// MSI Token Provider obtains a token from the MSI endpoint
+// The args parameter is a dictionary conforming with the ManagedIdentityCredential initializer API arguments
+export class MsiTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    clientId: string;
+    managedIdentityCredential!: ManagedIdentityCredential;
+
+    constructor(kustoUri: string, clientId: string) {
+        super(kustoUri);
+        this.clientId = clientId;
+    }
+
+    init(): void {
+        return;
+    }
+
+    async acquireToken(): Promise<TokenResponse> {
+        if (this.managedIdentityCredential == null) {
+            this.managedIdentityCredential = new ManagedIdentityCredential(this.clientId);
```

Comment:
Commonly this is an optional parameter and when not provided the SystemMI is accessed.
Is this the case with Node ManagedIdentityCredential?
If so we should support this

Reply:
Nope, you must send it in Node

Reply:
Is there a special notation for System (e.g, passing "system" as the oid)?

Reply:
Oops, I see now that indeed there is another ctor. I'll change clientId to optional